### PR TITLE
8271014: Refactor HeapShared::is_archived_object()

### PR DESCRIPTION
**Pull request by iklam**
(This PR is in preparation of [JDK-8270489](https://bugs.openjdk.java.net/browse/JDK-8270489) - Support archived heap objects in EpsilonGC)

We have the following function that's used ONLY in asserts and WhiteBox tests.

```
bool HeapShared::is_archived_object(oop p) {
  return Universe::heap()->is_archived_object(p);
}
```

This function is used for two completely different purposes:

[1] During dump time, is `p` an object that's about to be written into the archived heap?
[2] During runtime, is `p` an object in an archive region?

To support archived objects in other types of GCs that do not support archive regions, we should refactor this code to distinguish between the two use cases. This will make the code easier to understand.

=========

[0] Remove HeapShared::is_archived_object()
[1] -> add new function `HeapShared::is_archived_object_during_dumptime()`
[2] -> rewrite the code to call `Universe::heap()->is_archived_object(p)` instead.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271014](https://bugs.openjdk.java.net/browse/JDK-8271014): Refactor HeapShared::is_archived_object()

### Reviewers
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**)
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**) ⚠️ Review applies to a03a7ab306d60f3fdd22542eaa5a1030119225fc

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4855/head:pull/4855` \
`$ git checkout pull/4855`

Update a local copy of the PR: \
`$ git checkout pull/4855` \
`$ git pull https://git.openjdk.java.net/jdk pull/4855/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4855`

View PR using the GUI difftool: \
`$ git pr show -t 4855`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4855.diff">https://git.openjdk.java.net/jdk/pull/4855.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/cds/archiveUtils.cpp [hunk 1]
```diff
@@ -39,6 +39,7 @@
 #include "oops/compressedOops.inline.hpp"
 #include "runtime/arguments.hpp"
 #include "utilities/bitMap.inline.hpp"
+#include "utilities/formatBuffer.hpp"
```
Comment (calvinccheung) [inline src/hotspot/share/cds/archiveUtils.cpp:42]: Why add the above include?

File: src/hotspot/share/cds/heapShared.cpp [hunk 2]
```diff
@@ -116,6 +116,14 @@ GrowableArrayCHeap<oop, mtClassShared>* HeapShared::_pending_roots = NULL;
 narrowOop HeapShared::_roots_narrow;
 OopHandle HeapShared::_roots;
 
+#ifdef ASSERT
+bool HeapShared::is_archived_object_during_dumptime(oop p) {
+  assert(HeapShared::is_heap_object_archiving_allowed(), "must be");
+  assert(DumpSharedSpaces, "must be called during dump time only");
```
Comment (calvinccheung) [inline src/hotspot/share/cds/heapShared.cpp:122]: I'd suggest to change it to the following to be consistent with asserts in other places:
`assert(DumpSharedSpaces, "this function is only used with -Xshare:dump");`

**Review by calvinccheung — APPROVED** [review]
Looks good. Just one question and one suggestion.

**Review by yminqi — APPROVED** [review]
LGTM.

File: src/hotspot/share/cds/archiveUtils.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/archiveUtils.cpp:42]: archiveUtils.cpp uses `err_msg`, which is declared in formatBuffer.hpp. Before this fix, this header was included indirectly from heapShared.inline.hpp -> collectedHeap.hpp -> formatBuffer.hpp.

In this PR, I removed the inclusion of collectedHeap.hpp in heapShared.inline.hpp, so the missing dependency was exposed.

File: src/hotspot/share/cds/heapShared.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/cds/heapShared.cpp]: Fixed.

**Review by calvinccheung — APPROVED** [review]

**Comment by iklam** [comment]
Thanks @calvinccheung and @yminqi for the review.
/integrate

**Pull request closed without merging**